packages/sources/universal-v3/src/utils/LambdaController.ts

```diff
@@ -0,0 +1,348 @@
+  source: string
+  requestId: string
+  secrets: Record<string, unknown>
+  args?: string[]
+  numAllowedQueries?: number
+  maxResponseBytes?: number
+}
+
+type TimeLastUsed = number
+
+interface DeployedLambdaFunctions {
+  timeLastUsed: Record<string, TimeLastUsed>
+  count: number
+}
+
+export class LambdaController {
+  private deployedLambdaFunctions?: DeployedLambdaFunctions
+  private initalized = false
+  private lambda: AWS.Lambda
+  private logger = makeLogger('LambdaController')
+
+  constructor(private config: AdapterConfig<typeof customSettings>) {
+    this.lambda = new AWS.Lambda({ apiVersion: '2015-03-31' })
+  }
+
+  public initalize = async (): Promise<void> => {
+    if (this.initalized) {
+      throw Error('LambdaController already initalized')
+    }
+    AWS.config.update({
+      region: process.env['AWS_REGION'] as string,
+      credentials: {
+        accessKeyId: this.config.AWS_ACCESS_KEY_ID,
+        secretAccessKey: this.config.AWS_SECRET_ACCESS_KEY,
+      },
+    })
+    this.deployedLambdaFunctions = await this.getDeployedLambdaFunctions()
+    this.logger.debug(
+      `${this.deployedLambdaFunctions?.count} Deployed Lambda functions: ${Object.keys(
+        this.deployedLambdaFunctions.timeLastUsed,
+      )}`,
+    )
+    this.initalized = true
+  }
+
+  // This background tasks monitors all deployed Lambda functions used by the backgroundExecuter of the EAv3 framework.
+  // When LAMBDA_MAX_DEPLOYED_FUNCTIONS is reached, the least recently used function will be deleted.
+  // The returned number is how many milliseconds to wait before the next execution.
+  public lambdaPruner = async (): Promise<number> => {
+    if (!this.deployedLambdaFunctions) {
+      throw Error('LambdaController class not initalized')
+    }
+
+    if (this.deployedLambdaFunctions.count < this.config.LAMBDA_MAX_DEPLOYED_FUNCTIONS) {
+      return this.config.LAMBDA_PRUNER_LOOP_WAIT_MS
+    }
+
+    const leastRecentlyUsedFunction = await this.getLeastRecentlyUsedFunction()
+
+    try {
+      await this.deleteLambdaFunction(leastRecentlyUsedFunction)
+      delete this.deployedLambdaFunctions.timeLastUsed[leastRecentlyUsedFunction]
+      this.deployedLambdaFunctions.count--
+    } catch (error) {
+      this.logger.error(`Error deleting lambda function ${leastRecentlyUsedFunction}: ${error}`)
+      return 1
+    }
+
+    return 1
+  }
+
+  public executeRequestInLambda = (
+    FunctionName: string,
+    sandboxRequestData: SandboxRequestData,
+    retryCount: number,
+  ): Promise<SandboxOutput> => {
+    if (!this.deployedLambdaFunctions) {
+      throw Error('LambdaController not initalized')
+    }
+
+    const invokeRequest: Lambda.InvocationRequest = {
+      FunctionName,
+      Payload: JSON.stringify(sandboxRequestData),
+    }
+
+    return new Promise((resolve, reject) => {
+      this.lambda.invoke(invokeRequest, (err, data) => {
+        this.logger.trace(`Invoking Lambda function name ${FunctionName}`)
```

Comment:
It does not unfortunately AFAIK

Reply:
https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Lambda.html#createFunction-property

In v2 of the aws-sdk you can do it by not providing a callback and chaining .promise() on the returned request, i.e.
```typescript
this.lambda.invoke(invokeRequest).promise()
```
It would be better to migrate to v3 of the sdk since it's the current one and v2 will be deprecated (or is now, I'm not sure), but I think v3 is built for Node 18 so there might be some additional work needed, we don't want to run EAs on different node versions